File: backend/app/schemas/query.py
from pydantic import BaseModel, Field


class QueryProcessRequest(BaseModel):
    query: str = Field(..., min_length=1)


class SQLValidationPayload(BaseModel):
    valid: bool
    errors: list[str] = Field(default_factory=list)


class QueryProcessResponse(BaseModel):
    generation_mode: str
    generated_sql: str
    corrected_sql: str | None = None
    executed_sql: str | None = None
    validation: SQLValidationPayload
    cache_hit: bool
    similarity_score: float
    validation_status: str
    validation_errors: list[str] = Field(default_factory=list)
    execution_time: float
    rows_returned: int
    results: list[dict[str, str | int | float | bool | None]]


class LLMSQLGenerateRequest(BaseModel):
    query: str = Field(..., min_length=1)


class LLMSQLGenerateResponse(BaseModel):
    generated_sql: str
    generation_mode: str

File: frontend/src/api/query.ts
import { apiPost } from './client';
import { normalizeGenerationMode, type GenerationMode } from '../utils/generationMode';

export type QueryProcessRequest = {
  query: string;
};

export type QueryResultRow = Record<string, string | number | boolean | null>;

export type QueryProcessResponse = {
  generation_mode: GenerationMode;
  generated_sql: string;
  corrected_sql: string | null;
  executed_sql: string | null;
  validation: {
    valid: boolean;
    errors: string[];
  };
  cache_hit: boolean;
  similarity_score: number;
  validation_status: 'valid' | 'invalid';
  validation_errors: string[];
  execution_time: number;
  rows_returned: number;
  results: QueryResultRow[];
};

export async function processQuery(query: string) {
  const response = await apiPost<QueryProcessRequest, QueryProcessResponse>('/query/process', { query });
  return {
    ...response,
    generation_mode: normalizeGenerationMode(response.generation_mode)
  };
}
